LineNumbers.md

```diff
@@ -0,0 +1,320 @@
+
+Most languages have linenumber directives. They are used for generated code so
+debuggers and errormessages are able to tell which line from the source file is
+responsible instead of the line from the intermediate representation. Golang
+uses '//line [filename]:[linenumber]', where [XXX] denotes an optional
+parameter. Many (most?) languages from the C-family instead used the format
+'#line lineno ["filename"]'.
+
+This is quite usefule, since `go build`, `go run` etc will provide references
+to the markdown file with the bug (well, at least in theory). IF we want to
+debug the intermediate file, a reference above the specific code is always
+readily avialible a quick right click (Acme) or `gF` away (vim).
+
+If we want to implement this for lmt, we need to change quite a few things, the
+most obvious is the internal representation of a line, which needs quite a lot
+of more metadata, as the filename, the linenumber and the language of the
+codeblock.
+
+We remake CodeBlock to a slice of CodeLines. For every line we don't only
+record the text but the name of the source file, the linenumber and the
+language (for different line directives).
+
+```go "global block variables"
+type File string
+type CodeBlock []CodeLine
+type BlockName string
+type language string
+type CodeLine struct {
+	text   string
+	file   File
+	lang   language
+	number int
+}
+
+var blocks map[BlockName]CodeBlock
+var files map[File]CodeBlock
+```
+
+Handling a block line is now not as simple as just joining two strings. We
+start appending CodeLines into CodeBlocks.
+
+```go "Handle block line"
+block = append(block, line)
+```
+
+This cascades. The CodeBlocks in the maps files and blocks can't be joined, but
+has to be appended. We take a few moments to clean up the emptying of metadata
+per line, i.e. if a block is ending we can unset inBlock.
+
+```go "Handle block ending"
+inBlock = false
+// Update the files map if it's a file.
+if fname != "" {
+	if appending {
+		files[fname] = append(files[fname], block...)
+	} else {
+		files[fname] = block
+	}
+}
+
+// Update the named block map if it's a named block.
+if bname != "" {
+	if appending {
+		blocks[bname] = append(blocks[bname], block...)
+	} else {
+		blocks[bname] = block
+	}
+}
+```
+
+To get this to work we need to start using CodeLine as line in our
+implementation of the file processing. We also should to save the filename
+directly when opening the file. To simplify we break out our variables from the
+implementation of file processing.
+
+```go "process file implementation variables"
+scanner := bufio.NewReader(r)
+var err error
+
+var line CodeLine
+line.file = File(inputfilename)
+
+var inBlock, appending bool
+var bname BlockName
+var fname File
+var block CodeBlock
+```
+
+When processing our files we need to increment the file line counter for every
+line read.
+
+```go "process file implementation"
+<<<process file implementation variables>>>
+for {
+	line.text, err = scanner.ReadString('\n')
```

Comment:
not sure if it's just how GitHub is showing it or not, but the indentation here looks wrong. It should be tabs. You can highlight the code block in ie. acme and pipe that section through `gofmt` to make sure the formatting of theGo code is formatted correctly. (If there are <<< .... >>> macros as a hack you can temporarily comment them, use gofmt, then uncomment them)